### PR TITLE
Added support for junit2trx

### LINKED ISSUE #60: JUnit2Trx
Is there a tool that converts test results in the opposite way? It would be valuable for Azure DevOps, where some functionality only works with **trx** results

Can your tool be extended to do this conversion?

### PR DESCRIPTION
**Pull request by gfoidl**
Fixes https://github.com/gfoidl/trx2junit/issues/60

Refactored `ITestConverter` out from the `...Builder`, moved around some code to form an abstraction that allows easier modes for `trx2junit` and `junit2trx`.

From user point of view a command line option `--junit2trx` got added, that performs the conversion of junit-files to trx-files.

### CODE REVIEW

**Comment by gfoidl** [comment]
@yanlend here is my spike for _junit2trx_-conversion. A [preview version](https://www.myget.org/feed/gfoidl/package/nuget/trx2junit) can be installed with

```shell
# deinstall current version
dotnet tool uninstall -g trx2junit

# install preview
dotnet tool install -g trx2junit --version 1.3.0-preview-1 --add-source https://www.myget.org/F/gfoidl/api/v3/index.json 
```

_trx_ is quite cumbersome (junit is a lot leaner and cleaner), so I hope that I got enough quirks :wink:
Can you give it a try?

**Comment by yanlend** [comment]
It does not seem to recognize the `--junit2trx` command as an option in the preview1, but as a file.

**Comment by gfoidl** [comment]
Hm, I hit this in some tests too (especially in docker), in others it recognized it. Will investigate that.

**Comment by yanlend** [comment]
Ah you fixed the bug, I found it as well. Great!

**Comment by gfoidl** [comment]
Oh, that was a dumb oversight by me, and not covered by tests properly. It's fixed now (as you've seen).

Please give it another try after CI is ready and the new preview available.
```shell
# deinstall current version
dotnet tool uninstall -g trx2junit

# install preview
dotnet tool install -g trx2junit --version 1.3.0-preview-2 --add-source https://www.myget.org/F/gfoidl/api/v3/index.json 
```
(myget needs a couple of minutes until it can serve this version).

**Comment by yanlend** [comment]
Ok this part is fixed, thanks a lot. Now the XML the behave tool produces is not valid according to the schema you are using. It works with the current [jenkins schema](https://raw.githubusercontent.com/junit-team/junit5/master/platform-tests/src/test/resources/jenkins-junit.xsd)

**Comment by gfoidl** [comment]
OK. I'll make the parser tolerant, so that both schemas can be read.

Do you have an example xml that follows the jenkins schema? So I can use that for the tests (and check it in)?

**Comment by yanlend** [comment]
I attached an output file but had to rename it to .txt to be able to upload it.
[TESTS-PatientRegistration.txt](https://github.com/gfoidl/trx2junit/files/3842321/TESTS-PatientRegistration.txt)

**Comment by gfoidl** [comment]
Thanks. I'll look into this a bit later today.

File: source/trx2junit/Internal/junit2trx/TrxTestResultXmlBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Xml.Linq;
+using trx2junit.Models;
+
+namespace trx2junit
+{
+    public class TrxTestResultXmlBuilder : TrxBase, ITestResultXmlBuilder<TrxTest>
+    {
+        private static readonly Guid s_testTypeId = Guid.Parse("13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b");
```
Comment (gfoidl) [inline source/trx2junit/Internal/junit2trx/TrxTestResultXmlBuilder.cs:9]: Visual Studio can open trx-files, but not individual tests. So this exact guid has to be used, in order to open the tests.
Found this guid by looking at other _trx_-files.

Maybe there are some other distinct guid-values. I don't know. If there is any special value, please let me know.

**Comment by gfoidl** [comment]
When CI and myget are ready, [preview-3](https://www.myget.org/feed/gfoidl/package/nuget/trx2junit/1.3.0-preview-3) is ready to test :wink:

In CI I've added a test for the jenkins-style https://github.com/gfoidl/trx2junit/blob/897f58c4f7ad3c7d6a528e1c12b04cf0efecb7a9/tests/scripts/run-junit2trx.sh#L25-L30 and that passes. Hopefully it works for you too.

File: tests/trx2junit.Tests/data/junit/yanlend.xml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+<testsuite errors="1" failures="0" hostname="samplename" name="PatientRegistration.Patient Registration (Example)" skipped="0" tests="3" time="0.036988" timestamp="2019-11-08T09:10:24.803951">
```
Comment (gfoidl) [inline tests/trx2junit.Tests/data/junit/yanlend.xml]: I should rename this file to jenkins-style.xml or similar.

**Comment by yanlend** [comment]
Awesome, it works. I can convert to a valid trx file. Thank you very much @gfoidl

**Comment by gfoidl** [comment]
Perfect! Thanks for the input -- which led to quite some refactoring in the code-base.

Later today I'm going to merge this PR, then I have a minor polish for some allocations in the code. After that I'll release `v1.3.0`. In the meantime you can use the preview available...

File: ReadMe.md [hunk 3]
```diff
@@ -41,6 +43,12 @@ $ trx2junit a.trx --output ../results
 $ trx2junit --output results a.trx ../tests/b.trx
 ```
 
+### junit to trx
+
+With option `--junit2trx` a conversion from _junit_ to _trx_ can be performed.
+
+If a given _xml_-file is not a junit-file, a message will be logged to _stderr_ and the exit-code is set to 1.
```
Comment (gfoidl) [inline ReadMe.md]: Add a sentence about the schemas that are understood by the program.

**Pull request merged**